sae_lens/training/training_sae.py

```diff
@@ -225,17 +277,51 @@ def training_forward_pass(
         else:
             ghost_grad_loss = 0.0
 
-        # SPARSITY LOSS
-        # either the W_dec norms are 1 and this won't do anything or they are not 1
-        # and we're using their norm in the loss function.
-        weighted_feature_acts = feature_acts * self.W_dec.norm(dim=1)
-        sparsity = weighted_feature_acts.norm(
-            p=self.cfg.lp_norm, dim=-1
-        )  # sum over the feature dimension
+        if self.cfg.architecture == "gated":
+            # Gated SAE Loss Calculation
+
+            # original, deprecated gated SAE L1 loss calculation            
+            # via_gate_feature_magnitudes = torch.relu(
+            #     sae_in_centered @ self.W_enc + self.b_gate
+            # )
+            # preactivation_l1_loss = (
+            #     current_l1_coefficient
+            #     * torch.sum(via_gate_feature_magnitudes, dim=-1).mean()
+            # )
+            
+            # Shared variables
+            sae_in_centered = sae_in - self.b_dec
+            pi_gate = sae_in_centered @ self.W_enc + self.b_gate
+            
+            # SFN sparsity loss - summed over the feature dimension and averaged over the batch
+            sfn_sparsity_loss = current_l1_coefficient * torch.sum(
+                torch.abs(pi_gate) * self.W_dec.norm(dim=1), dim=-1
```

Comment:
This is incorrect, we should apply the Relu then use pi_gate_act here. Fixing this results in working benchmark runs with/without SFN.